### ISSUE TITLE
Blank Categories and Author on Custom Dimensions tiles

### ISSUE THREAD

**Issue by bethanylang** [google/site-kit-wp#7755]
## Bug Description

As reported by @wpdarren in Bug Bash ([Asana task](https://app.asana.com/0/1205508124049666/1205767360235585/f)):

We noticed that some categories are blank on the custom dimension tiles even though they shouldn't be.   On my live site I have at least one category and author against a post so wouldn't expect to see any blank data. 

![image](https://github.com/google/site-kit-wp/assets/12189958/6b5abf38-a6ce-401f-a7e7-4786620c399d)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

Please see [this comment](https://github.com/google/site-kit-wp/issues/7755#issuecomment-1787161277) when reviewing these AC.

- When tracking the `googlesitekit_post_categories` custom dimension, if an error occurs when retrieving the list of category IDs, the custom dimension should not be recorded at all.
- This will have the effect of adding to the `(not set)` category when viewing the "Top categories by pageviews" Key Metric widget.

## Implementation Brief

Within `includes/Modules/Analytics_4.php`:
- When looping over the `availableCustomDimensions` settings in `get_custom_dimensions_data`:
  - Use `get_the_category` rather than `wp_get_post_categories` to retrieve the list of categories.
  - This means we can remove the explicit error handling, see https://github.com/google/site-kit-wp/issues/7755#issuecomment-1795061474
  - In the case where there are no categories _do not_ set the `googlesitekit_post_categories` custom dimension in `$data`.

### Test Coverage

- Add test coverage for the above scenario (see [`Analytics_4Test::test_get_custom_dimensions_data`](https://github.com/google/site-kit-wp/blob/6d6bbff14441dabc1391ccc90b246bbbab9cda5a/tests/phpunit/integration/Modules/Analytics_4Test.php#L2834)).

## QA Brief

The steps to reproduce for this issue are not clear, so effectively the best bet is to ensure that the categories appear and no instances of empty categories appear in the "Top categories by pageviews" widget.

* Set up Site Kit with the `keyMetrics` feature flag enabled
* Ensure the "Top categories by pageviews" does not have any empty entries. Instead, any empty entries should read "(not set)"

**Note:** This change will not remove empty entries on old sites, because it was caused by submitting a custom dimension without categories set, so please don't QA this by checking an existing site with empty entries and seeing if this fixes it 🙂 

## Changelog entry

* Avoid tracking a blank category for the `googlesitekit_post_categories` custom dimension.

**Comment by techanvil** [comment]
Please note that I've spent some time investigating this in IB before moving back to AC having determined there is a validly [reachable case](https://github.com/google/site-kit-wp/blob/7e36531c945151f051d44ae37186fa685090ad85/includes/Modules/Analytics_4.php#L1404-L1406) where the category ID may be tracked as an empty string.

Beyond this, it's worth noting that the issue as originally [reported](https://app.asana.com/0/1205508124049666/1205767360235585/f) by @wpdarren is no longer evident on his live site since updating to the latest `develop`. I've further investigated and found that there are no categories or authors that have been tracked for that site's property with an empty ID, so it looks likely the update to `develop` has brought in a fix.

However @wpdarren did also mention he's potentially seen a blank category on `oi.ie` too - taking a look at the reported data for this site's property it's evident there are empty IDs as seen here. However with `oi.ie` being used heavily for development I think it's quite likely these have been introduced through previous iterations and are not necessary representative of the current state of the codebase.

![image](https://github.com/google/site-kit-wp/assets/18395600/30204026-4986-485b-9d8d-86df91253193)

So, in conclusion, having realised there _is_ an unlikely but possible scenario for an empty category ID, I've added a point to the AC to address this and have moved it to ACR to get a review of the suggested solution.

Please feel free to assign this back to me once this returns to the IB column.

**Comment by tofumatt** [comment]
This approach sounds good, assigning it to @techanvil for the IB as requested 🙂

**Comment by aaemnnosttv** [comment]
@techanvil regarding blank categories, we probably shouldn't be sending the parameter at all if it resulted in an empty string https://github.com/google/site-kit-wp/blob/9054bd58aead576ebf25b1863260753277528b59/includes/Modules/Analytics_4.php#L1406

The other custom taxonomies should never have that problem, so I think the blank value is probably only possible for categories.

It seems like we should avoid sending this to GA to begin with. If the category taxonomy didn't exist (is this the only way this is possible?), or wasn't registered for Posts (unlikely edge case) we should probably surface this in another way, or prevent that tile from being chosen but this is probably out of scope for this issue.

**Comment by techanvil** [comment]
As per the above:

It's worth noting that there hasn't actually been a report of the author name being blank, so we should not assume it will be unless we see it happening on a live site. So, we can focus on the behaviour for categories here.

While reviewing the code, the only identified case where the category can be recorded and presented as an empty string is when an error occurs while retrieving the category IDs to track for the custom dimension, and the CD is then tracked with an empty string.

It should however further be noted that this is a very unlikely edge case and highly unlikely to be the cause of the issue as initially reported by @wpdarren. As @aaemnnosttv has alluded to in the related [Asana task](https://app.asana.com/0/0/1205767360235585/1205788136098478/f), the empty category Darren's seen could be due to the custom dimension being tracked with an empty string or similar in a previous version of Site Kit while the feature was still in development. While Darren has [re-reported](https://10up.slack.com/archives/CBKKQEBR9/p1698642000776659) seeing the issue, it was no longer reproducible when I was available to take a look so we don't have a clear picture of what the cause for this is and if it's relevant to the code as it stands in `develop`.

Therefore, all we can reasonably do at this stage is to address the identified case where the category can be recorded as an empty string, and then keep an eye out for any further cases where the category is seen as an empty string on a live site. If that occurs we can dig in to identify the cause.

**Comment by eclarke1** [comment]
@aaemnnosttv are you able to pick up the IB review here please to get it into EB

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, that's a good shout. I've updated the IB accordingly.

**Comment by aaemnnosttv** [comment]
Thanks @techanvil 👍 

IB ✅

**Comment by techanvil** [comment]
Hey @tofumatt, it looks like there's a bit of a misunderstanding in the QAB with regard to how this will affect `(not set)` entries that I'm guessing stems from this [comment](https://github.com/google/site-kit-wp/pull/7834#pullrequestreview-1717849208) in the CR.

Rather than eliminating `(not set)` entries, this change should actually have the effect of _adding_ a `(not set)` entry any time the condition is hit and the `googlesitekit_post_categories` custom dimension is not tracked for the given `post` page impression. This is briefly mentioned in the AC.

What it's avoiding is the blank category entry as per the screenshot in the feature description. GA4 is still tracking an event, so it will presumably be registering the custom dimension as `(not set)` as [no value is tracked for it](https://support.google.com/analytics/answer/13504892?hl=en#zippy=%2Cin-this-article).

I've merged the code, but please can you update the QAB accordingly?

Cc @zutigrm

**Comment by techanvil** [comment]
Further on the above, it's worth noting that we might be entirely [filtering out](https://github.com/google/site-kit-wp/issues/7737#issuecomment-1801643796) `(not set)` entries from the "Top categories" widget in future, so the QAB should probably just focus on the removal of blank categories aspect...

**Comment by tofumatt** [comment]
@techanvil Oh, right, okay! I couldn't replicate either so I wasn't 100% there, thanks 👍🏻 

I've updated the QA Brief 👍🏻

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

**Verified:**

- I have tested this on a few live sites, and no blank categories or authors appear. 
- Set up [a test site](https://key-metrics-test.s2-tastewp.com/) and will see what that looks like when I log back in on Monday.

**Comment by wpdarren** [comment]
QA Update: ✅ 

**Verified:**

- I tested this on a few live sites, including a brand-new site with custom dimensions tracked for categories and authors, and found no empty data. 

![image](https://github.com/google/site-kit-wp/assets/73545194/c9eb984a-4c5b-473b-9aaa-e848961b1f78)
